Civi/Token/TokenCompatSubscriber.php

```diff
@@ -56,7 +56,10 @@ public function setupSmartyAliases(TokenValueEvent $e) {
    */
   public function onRender(TokenRenderEvent $e): void {
     $useSmarty = !empty($e->context['smarty']);
-    $e->string = $e->getTokenProcessor()->visitTokens($e->string, function() {
+    $e->string = $e->getTokenProcessor()->visitTokens($e->string, function($token = NULL, $entity = NULL, $field = NULL, $filterParams = NULL) {
```

Comment:
without CiviContribute there were issues in the event online receipt (and other related templates) because they would have tokens like 

{contribution.tax_amount|boolean} in if statements and this was filtering that down to an empty string which then if there were multiple parts to the if e.g. {if $showTaxAmount && {contribution.tax_amount|boolean}} it was being passed to smarty as {if $showTaxAmount && } which of course is a PHP syntax error this causes it to resolve any that use boolean filter to be 0

Reply:
OK - let me see if I can move this into where booleans are otherwise rendered - ie this sort of area

https://github.com/civicrm/civicrm-core/blob/1ce5c9bf6b5d93e4ffd6bc2f44d5226aeb1141d2/Civi/Token/TokenProcessor.php#L494-L499

Reply:
So the problem is we don't get to there because of https://github.com/civicrm/civicrm-core/blob/1ce5c9bf6b5d93e4ffd6bc2f44d5226aeb1141d2/Civi/Token/TokenProcessor.php#L381

Reply:
@seamuslee001 yeah we could do 

![image](https://github.com/civicrm/civicrm-core/assets/336308/e27ff8aa-2797-4fee-9770-7fd00b2f6bf7)


Which feels more right to me - in that generally token filters are handled in the token system. OTOH I am somewhat terrified of the code in `vistiTokens` and associated functions cos it's hard to follow which might be a case for not touching it

Reply:
yeh that might be right but yeh likewise and it seemed like as this was doing the "cleanup" it might be the most sensible place to put things

Reply:
ok - well I'm otherwise happy with this PR so @totten can merge this as is or opt for the other location to handle unset booleans

Reply:
OK, this is the first time I've really looked at `|boolean`. So I'm trying to catch-up. Here's what I'm understanding:

* The idea in 6543dfe601b0c5ac58b2062d8fe5d409c1a9d313 is to help write Smarty conditionals that key-off of token-data, eg
    ```smarty
    {if {contact.first_name|boolean}}...{/if}
    ```
* This kind of `{if {token}}` notation is the quintessential quirk of Token-Smarty format. It's why Token-Smarty is evil and should never be used. It's also why we have to keep it around -- because changing to a simpler/singular language would break these conditionals.
* So I guess the idea of `|boolean` is to make the conditional a bit safer, eg
    ```smarty
    {if {contact.first_name}}...{/if}        {* More evil *}
    {if {contact.first_name|boolean}}...{/if} {* Less evil *}
    ```
* In the callpath of this specific bug, we have a CiviEvent message with expression:
    ```smarty
    {if $isShowTax && {contribution.tax_amount|boolean}}
    ```
* In CiviEvent contexts, the token `{contribution.tax_amount}` may or may not be defined. (*CiviEvent is only softly dependent on CiviContribute.*)
* The approach here is to define `{any_unkn.own|boolean}` to output `0`.
* The question you're mulling is: Where, exactly, should we put in the override to make `{any_unkn.own|boolean}` evaluate to `0`?

Reply:
So, yes, to my eye, both locations work, and the `TokenProcessor::$getToken()`  seems a bit more poignant than `TokenCompatSubscriber::onRender()`.

Whichever location it's in, I think it should have a code-comment explaining why it's there, e.g.

```php
if ($filterParams && $filterParams[0] === 'boolean') {
  // This token was missed during primary rendering, and it's supposed to be coerced to boolean.
  // Treat an unknown token as false-y.
  return 0;
}
```

Reply:
@totten that seems about right

Reply:
Force-pushed a commit with the extra comment.